lms/add-teacher-notifications

anathomical opened the pull request:
## WHAT
Add a model to store teacher notification records in
## WHY
We want to establish an abstraction to store messages in that can be used to enqueue messages to teachers and record their various delivery statuses.  This doesn't really do anything yet, but is foundational for further additions.
## HOW
Add a new table and model to store notification data in.  Add some strict validation on the `params` JSONB value so that it has a known and consistent shape.

### Notion Card Links
https://www.notion.so/quill/Send-a-roll-up-email-of-student-related-events-to-teachers-at-a-frequency-of-their-choosing-a074aabc41eb41be84f732db7a63aa49?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | No because this is foundational, and doesn't contain actual new functionality
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Add a model to store teacher notification records in

anathomical committed:
Use a datetime for email_sent instead of a boolean

anathomical committed:
Fix linting errors

anathomical committed:
Add association from user to teacher_notification

anathomical committed:
Make teacher_notifications dependent destroy on users

brendanshean commented on services/QuillLMS/app/models/teacher_notification.rb:
```diff
@@ -0,0 +1,68 @@
+  validate :params_must_be_hash, :params_payload_has_required_keys
+
+  private def params_is_hash?
+    params.is_a?(Hash)
```
Since `params` has special meaning in rails in the context of the controller, to avoid confusion later, I would name this something more descriptive (e.g. custom_attrs).

brendanshean commented on services/QuillLMS/spec/factories/teacher_notification.rb:
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :teacher_notification, class: TeacherNotification do
```
I think the  `class: TeacherNotification` is unnecessary.

brendanshean commented on services/QuillLMS/spec/factories/teacher_notification.rb:
```diff
@@ -0,0 +1,14 @@
+
+FactoryBot.define do
+  factory :teacher_notification, class: TeacherNotification do
+    association :user
```
I think you can just use `user` here.

brendanshean commented on services/QuillLMS/spec/models/teacher_notification_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    it 'should require params to be a hash' do
+      teacher_notification.params = ''
+
+      expect(teacher_notification.valid?).to be(false)
```
All four of the `.valid?` can be written using the `be_valid` matcher, for example:

expect(teacher_notification).to be_valid

happythenewsad commented on services/QuillLMS/app/models/teacher_notification.rb:
```diff
@@ -0,0 +1,68 @@
+  private def params_payload_has_required_keys
+    return unless params_is_hash?
+
+    return if params.keys.map(&:to_s).to_set == NOTIFICATION_TYPE_VALIDATION[notification_type].map(&:to_s).to_set
```
Nit: set inclusion would be a bit more flexible than set equality here. We don't care if the payload has more stuff than necessary, but we do care if the payload is lacking a key.

anathomical replied:
@happythenewsad My thought on this was that we'll know exactly what keys we do want, and that additional keys will only be noise.  By doing equality instead of inclusion, my hope was to prevent bloat to these JSON blobs.  Having extra keys probably won't hurt anything other than potentially taking up space in the database, but what are your thoughts on preferring to allow extra keys instead of restricting them?

happythenewsad replied:
This reasoning makes sense to me - carry on.

I was thinking that it would be annoying to have specs fail when adding a new expected key to the payload, but that's more of a feature than a bug, given that we've suffered from heterogenous jsonb schemata in the past.

anathomical replied:
I ended up doing a refactor based on @dandrabik 's feedback, which has allowed us to avoid a custom validator of this type.  Does this address your concerns here?

happythenewsad reviewed (approved):
Good stuff

anathomical committed:
Address PR feedback

dandrabik commented on services/QuillLMS/app/models/teacher_notification.rb:
```diff
@@ -0,0 +1,68 @@
+    STUDENT_COMPLETED_ALL_ASSIGNED_ACTIVITIES = 'student-completed-all-assigned-activities'
+  ]
+
+  NOTIFICATION_TYPE_VALIDATION = {
```
So I'm not opposed to this setup, but some Rails-y things to consider using instead. 

Consider using [store](https://api.rubyonrails.org/classes/ActiveRecord/Store.html). It's a way to store the json, but treat them as proper attributes. You could then use standard Rails validations, e.g. `validates :student_name, presence: true` would work. You can make validations conditional with `if:` clauses if needed. Also code such as `notification.student_name = "Thomas"` would work as expected.

anathomical replied:
I was originally struggling to figure out how to use `store` in conjunction with the fact that different notification types had different shapes, but it turns out that the solution was just to use the STI pattern.  So I've refactored the models in a way that I think takes advantage of both suggestions.  Let me know if this looks like a good solution.

dandrabik commented on services/QuillLMS/db/migrate/20230420141952_create_teacher_notifications.rb:
```diff
@@ -0,0 +1,14 @@
+class CreateTeacherNotifications < ActiveRecord::Migration[6.1]
+  def change
+    create_table :teacher_notifications do |t|
+      t.references :user, index: true, foreign_key: true, null: false
```
FK on users, we should deploy this during low traffic.

dandrabik commented on services/QuillLMS/app/models/teacher_notification.rb:
```diff
@@ -0,0 +1,68 @@
+#  fk_rails_...  (user_id => users.id)
+#
+class TeacherNotification < ApplicationRecord
+  NOTIFICATION_TYPES = [
```
Note, if there is different logic for each of these type, we could consider [STI](https://medium.com/@GeneHFang/single-table-inheritance-in-rails-6-emulating-oop-principles-in-relational-databases-be60c84e0126) here. This is something we could also migrate to later if it's helpful.

dandrabik reviewed (approved):
Thanks for putting this together. I've left some comments for your consideration, but overall this is approved. I do think you should consider using ActiveRecord store to make validations/usage simpler, but I'll leave that up to you.

emilia-friedberg commented on services/QuillLMS/app/models/teacher_notification.rb:
```diff
@@ -0,0 +1,68 @@
+#
+class TeacherNotification < ApplicationRecord
+  NOTIFICATION_TYPES = [
+    STUDENT_COMPLETED_DIAGNOSTIC = 'student-completed-diganostic',
```
Typo here - "diganostic".

emilia-friedberg reviewed (approved):
Typo but otherwise LGTM!

anathomical committed:
Refactor TeacherNotification into an STI model

This also allows us to use `store` to handle validation of JSON keys, which is much cleaner than what we were otherwise doing.

dandrabik reviewed (approved):
Looks good to me!

brendanshean commented on services/QuillLMS/app/models/teacher_notification/student_completed_all_diagnostic_recommendations.rb:
```diff
@@ -0,0 +1,28 @@
+#
+#  fk_rails_...  (user_id => users.id)
+#
+class TeacherNotification::StudentCompletedAllDiagnosticRecommendations < TeacherNotification
```
I just noticed there was this update with namespacing.  For some technical [reasons](https://rubystyle.guide/#namespace-definition), these should use the module definition over the shortened version

```
module TeacherNotification
  class StudentCompletedDiagnostic < ::TeacherNotification
    ..
   end
end
```

Make sure to do this for all subclasses.

anathomical committed:
Use cleaner module syntax for namespacing TeacherNotification subclasses

anathomical committed:
Fix linting errors
